### PR TITLE
Render hacker application on Portal dynamically

### PR DESCRIPTION
**Pull request by tdanielles**
## Description
Ticket: [https://www.notion.so/nwplus/PA-10-Create-component-for-each-question-type-in-Portal-7b4745d2744b45e98b71a6e00dcb8bfc?pvs=4](https://www.notion.so/nwplus/PA-10-Create-component-for-each-question-type-in-Portal-7b4745d2744b45e98b71a6e00dcb8bfc?pvs=4)

- fetch questions from firebase and render dynamically in Portal
- made each question type a separate component and refactored some code
- made validation dynamic (only checks for required questions and optional questions that are filled out by hackers)
- as of right now, legal name, school, major, country, and portfolio fields are marked required in hacker app template
- important note: firebase fields for the questions that aren't hardcoded are dependent on the options from CMS; right now it is set up to be the camel case of the options

## Testing
- active hackathon rn is hackcamp 2024
- questions should be filled out completely from CMS before attempting to check Portal
- after you save questions from CMS, delete your application instance from firebase and reload Portal
- pls pls pls hunt for 🐛🪲🐞

### CODE REVIEW

**Comment by martincai8** [comment]
goated holy

**Review by DonaldKLee — CHANGES_REQUESTED** [review]
Trisha how r u so cracked? Everything lgtm, just one question
What does portfolio do? i see the question, but no options to input an answer
image

**Comment by tdanielles** [comment]
> Trisha how r u so cracked? Everything lgtm, just one question What does portfolio do? i see the question, but no options to input an answer image

@DonaldKLee Ooo portfolio is for the resume, github, linkedin, and portfolio links! Idt its applicable to basic info so i didn't put the logic on there-- i think i'll leave what i have here on portal and edit on the cms side instead such that ppl can't add portfolio questions to the basic info section and ppl can't add major, school, name, and country questions to the skills section

**Comment by martincai8** [comment]
actually such an insane pr :0 

will take a closer look at the code soon but here are just a couple of things i noticed from playing around with the preview/cms!
- users(organizers) shouldn't be able to save the application when there's a question with no formInputField specified
- put in docs to (highly) recommend to not change question options once applications are released
- questionnaire section isn't in CMS/is still hardcoded in portal, was this intentional?

logic wise everything seems to work holy

File: yarn.lock [hunk 1]
```diff
@@ -4,28 +4,28 @@
 
 "@ampproject/remapping@^2.2.0":
   version "2.3.0"
-  resolved "https://registry.yarnpkg.com/@ampproject/remapping/-/remapping-2.3.0.tgz#ed441b6fa600072520ce18b43d2c8cc8caecc7f4"
+  resolved "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.3.0.tgz"
```
Comment (martincai8) [inline yarn.lock:7]: did you run npm install instead of yarn

Comment (tdanielles) [reply yarn.lock:7]: i swear i ran yarn install 😳 what should i do to fix it? delete and run yarn install?

Comment (martincai8) [reply yarn.lock]: fixed

**Review by martincai8 — COMMENTED** [review]
what's it like being megamind

File: src/components/ApplicationQuestions/MultipleChoice.jsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { Select } from '../Input'
+import { ErrorMessage } from '../Typography'
+
+const MultipleChoice = ({ refs, errors, formInputs, onChange, question }) => {
+  const options = question.other ? [...question.options, 'Other'] : question.options
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/components/ApplicationQuestions/MultipleChoice.jsx:5]: added in

File: src/containers/Application/Part1.jsx [hunk 3]
```diff
@@ -1,47 +1,31 @@
...
-  const { application, updateApplication, forceSave } = useHackerApplication()
+  const { application, updateApplication, forceSave, basicInfoQuestions } = useHackerApplication()
   const [, setLocation] = useLocation()
   const [errors, setErrors] = useState({})
   const [loading, setLoading] = useState(false)
+  const [questionsByOrder, setQuestionsByOrder] = useState([])
+
+  useEffect(() => {
+    const fetchQuestionsByOrder = async () => {
+      const questions = await getQuestionsByOrder(basicInfoQuestions)
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/containers/Application/Part1.jsx:19]: lol u right

File: src/containers/Application/Review.jsx [hunk 4]
```diff
@@ -37,7 +38,7 @@ const Review = () => {
   }
 
   const handleSubmit = async () => {
-    const allErrors = validateEntireForm(application)
+    const allErrors = await validateEntireForm(application, basicInfoQuestions, skillsQuestions)
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/containers/Application/Review.jsx:41]: u right again

File: src/utility/firebase.js [hunk 5]
```diff
@@ -210,3 +211,75 @@ export const updateProject = (author, projectId, data, dbHackathonName) => {
...
+        transformedOptions.other = false
+      }
+      initialTemplate.questionnaire[question.formInput] = transformedOptions
+    } else {
+      initialTemplate.questionnaire[question.formInput] = ''
+    }
+  })
+
+  return initialTemplate
+}
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/utility/firebase.js:285]: big brain 🧠

File: src/utility/HackerApplicationContext.jsx [hunk 6]
```diff
@@ -120,6 +126,25 @@ export function HackerApplicationProvider({ children }) {
...
+    }
+    const fetchQuestionnaireQuestions = async () => {
+      const appQuestions = await getHackerAppQuestions(dbHackathonName, 'Questionnaire')
+      setQuestionnaireQuestions(appQuestions)
+    }
+
+    fetchBasicInfoQuestions()
+    fetchSkillsQuestions()
+    fetchQuestionnaireQuestions()
+  }, [dbHackathonName])
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/utility/HackerApplicationContext.jsx:146]: lgtm, but fsr the promise.all isnt working so i just awaited each of them individually in the set

File: src/components/ApplicationForm/HackathonInfo.jsx [hunk 7]
```diff
@@ -1,55 +1,43 @@
...
+import parse, { domToReact } from 'html-react-parser'
 
 const HackathonInfo = () => {
+  const { dbHackathonName } = useHackathon()
+  const [title, setTitle] = useState('')
+  const [content, setContent] = useState('')
+
+  useEffect(() => {
+    const fetchQuestions = async () => {
+      const questions = await getHackerAppQuestions(dbHackathonName, 'Welcome')
```
[the remark this answers is not in the document]
Comment (tdanielles) [reply src/components/ApplicationForm/HackathonInfo.jsx:15]: would you be able to add that 🥲 might not have capacity this week

**Review by martincai8 — APPROVED** [review]
lgtm 🐐 ~~since the fields are now dynamic, we have to make sure the filters on admin for hacker info get fetched and applied dynamically too right? we can beg jason to do it~~

**Review by DonaldKLee — APPROVED** [review]
lgtm such a big PR LOL, great job

**Pull request merged**